flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecLookupJoin.java

```diff
@@ -125,45 +130,86 @@
         implements SingleTransformationTranslator<RowData> {
 
+    public static final String FIELD_NAME_JOIN_TYPE = "joinType";
+    public static final String FIELD_NAME_JOIN_CONDITION = "joinCondition";
+    public static final String FIELD_NAME_TEMPORAL_TABLE = "temporalTable";
+    public static final String FIELD_NAME_LOOKUP_KEYS = "lookupKeys";
+    public static final String FIELD_NAME_EXIST_CALC_ON_TEMPORAL_TABLE = "existCalcOnTemporalTable";
+    public static final String FIELD_NAME_CALC_ON_TEMPORAL_TABLE_OUTPUT_TYPE =
+            "calcOnTemporalTableOutputRowType";
+    public static final String FIELD_NAME_CALC_ON_TEMPORAL_TABLE_PROJECTIONS =
+            "calcOnTemporalTableProjections";
+    public static final String FIELD_NAME_CALC_ON_TEMPORAL_TABLE_CONDITION =
+            "calcOnTemporalTableCondition";
+
+    @JsonProperty(FIELD_NAME_JOIN_TYPE)
     private final FlinkJoinType joinType;
     /**
      * lookup keys: the key is index in dim table. the value is source of lookup key either constant
      * or field from right table.
      */
+    @JsonProperty(FIELD_NAME_LOOKUP_KEYS)
     private final Map<Integer, LookupJoinUtil.LookupKey> lookupKeys;
-    /** the reference of temporal table to look up. */
-    private final RelOptTable temporalTable;
-    /** calc performed on rows of temporal table before join. */
-    private final @Nullable RexProgram calcOnTemporalTable;
+
+    @JsonProperty(FIELD_NAME_TEMPORAL_TABLE)
+    private final TemporalTableSourceSpec temporalTableSourceSpec;
+
+    @JsonProperty(FIELD_NAME_EXIST_CALC_ON_TEMPORAL_TABLE)
+    private final boolean existCalcOnTemporalTable;
+
+    @JsonProperty(FIELD_NAME_CALC_ON_TEMPORAL_TABLE_OUTPUT_TYPE)
+    private final @Nullable RelDataType calcOnTemporalTableOutputRowType;
+
+    @JsonProperty(FIELD_NAME_CALC_ON_TEMPORAL_TABLE_PROJECTIONS)
+    private final @Nullable List<RexNode> calcOnTemporalTableProjections;
+
+    @JsonProperty(FIELD_NAME_CALC_ON_TEMPORAL_TABLE_CONDITION)
+    private final @Nullable RexNode calcOnTemporalTableCondition;
+
     /** join condition except equi-conditions extracted as lookup keys. */
+    @JsonProperty(FIELD_NAME_JOIN_CONDITION)
     private final @Nullable RexNode joinCondition;
 
     protected CommonExecLookupJoin(
             FlinkJoinType joinType,
             @Nullable RexNode joinCondition,
             // TODO: refactor this into TableSourceTable, once legacy TableSource is removed
-            RelOptTable temporalTable,
-            @Nullable RexProgram calcOnTemporalTable,
+            TemporalTableSourceSpec temporalTableSourceSpec,
             Map<Integer, LookupJoinUtil.LookupKey> lookupKeys,
-            InputProperty inputProperty,
+            boolean existCalcOnTemporalTable,
+            @Nullable RelDataType calcOnTemporalTableOutputRowType,
+            @Nullable List<RexNode> calcOnTemporalTableProjections,
+            @Nullable RexNode calcOnTemporalTableCondition,
```

Comment:
We can remove `calc` from the names, 
calcOnTemporalTableProjections -> projectionOnTemporalTable
calcOnTemporalTableCondition -> filterOnTemporalTable

and we can remove calcOnTemporalTableOutputRowType, and the type can be derived from calcOnTemporalTableProjections.